### PR TITLE
Fix compilation of musl-based native images.

### PR DESCRIPTION
**Pull request by gradinac**

### CODE REVIEW

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/c/libc/GLibc.java [hunk 1]
```diff
@@ -40,12 +40,12 @@ public void prepare(Path directory) {
     }
 
     @Override
-    public List<String> getCCompilerOptions() {
+    public List<String> getAdditionalQueryCodeCompilerOptions() {
         return new ArrayList<>();
```
Comment (olpaw) [inline substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/c/libc/GLibc.java:44]: Why not use `Collections.emptyList()` here and also below.
There is no value in allowing these return values to be mutable lists.

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/c/libc/MuslLibc.java [hunk 2]
```diff
@@ -64,24 +64,22 @@ public void prepare(Path directory) {
         setUpSpecFile(directory);
     }
 
+    @Override
+    public List<String> getAdditionalQueryCodeCompilerOptions() {
+        List<String> options = new ArrayList<>();
+        /* Avoid the dependency to muslc for builds cross compiling to muslc. */
+        options.add("--static");
+        return options;
```
Comment (olpaw) [inline substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/c/libc/MuslLibc.java:72]: Arrays.asList("--static")

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/c/libc/LibCBase.java [hunk 3]
```diff
@@ -24,18 +24,53 @@
...
+import org.graalvm.nativeimage.ImageSingletons;
+
 public interface LibCBase {
 
     String PATH_DEFAULT = "<default>";
 
+    static <T extends AnnotatedElement> boolean isProvidedInCurrentLibc(T element) {
+        LibCBase currentLibC = ImageSingletons.lookup(LibCBase.class);
+        Libc targetLibC = element.getAnnotation(Libc.class);
+        return targetLibC == null || Arrays.asList(targetLibC.value()).contains(currentLibC.getClass());
```
Comment (olpaw) [inline substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/c/libc/LibCBase.java:42]: Maybe add a comment comment that this policy causes the use of `@Libc` allows to tag parts of libraries that are exclusive to a specific Libc implementation.

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/c/libc/LibCBase.java [hunk 4]
```diff
@@ -24,18 +24,53 @@
...
+        Class<?> declaringClass = method.getDeclaringClass();
+        return isProvidedInCurrentLibc(declaringClass);
+    }
+
+    static boolean isTypeProvidedInCurrentLibc(Class<?> clazz) {
+        if (!isProvidedInCurrentLibc(clazz)) {
+            return false;
+        }
+        Class<?> enclosingClass = clazz.getEnclosingClass();
+        return enclosingClass == null || isProvidedInCurrentLibc(enclosingClass);
```
Comment (olpaw) [inline substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/c/libc/LibCBase.java:58]: This should work with arbitrarily nested classes/packages. See `com.oracle.svm.hosted.ImageClassLoader#handleClass` for reference.

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/c/libc/Libc.java [hunk 5]
```diff
@@ -30,7 +30,7 @@
 import java.lang.annotation.Target;
 
 @Retention(RetentionPolicy.RUNTIME)
-@Target({ElementType.METHOD})
+@Target({ElementType.METHOD, ElementType.TYPE})
```
Comment (olpaw) [inline substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/c/libc/Libc.java:33]: You might want to add some more target types. See `org.graalvm.nativeimage.Platforms`

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/c/libc/GLibc.java (hunk 1 above)
Comment (gradinac) [reply substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/c/libc/GLibc.java]: I agree, changing it to `Collections.emptyList()`.

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/c/libc/MuslLibc.java (hunk 2 above)
Comment (gradinac) [reply substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/c/libc/MuslLibc.java]: I agree, the code is also cleaner that way. I have changed it to `Collections.asSingletonList`.

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/c/libc/LibCBase.java (hunk 3 above)
Comment (gradinac) [reply substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/c/libc/LibCBase.java]: This is a good point, added a javadoc explaining how the check if a type is provided works.

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/c/libc/LibCBase.java (hunk 4 above)
Comment (gradinac) [reply substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/c/libc/LibCBase.java]: I agree, I have modified the code accordingly.

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/c/libc/Libc.java (hunk 5 above)
Comment (gradinac) [reply substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/c/libc/Libc.java]: I agree, added `ElementType.PACKAGE` to the list of target types.

**Comment by gradinac** [comment]
> Nice work! A few small improvements and your good.

Thank you! :)

**Comment by gradinac** [comment]
As the previous musl fix had been merged to master these NFI changes, I had created another PR with the above change: [link](https://github.com/oracle/graal/pull/2371). @gilles-duboscq I had changed the logic so it works as you have described it above (with `hasIsolatedNamespaces()`, it makes much more sense that way when I looked at it again. :)

The PR has since been merged, so I will close this PR.

**Pull request closed without merging**